File: vmsserver-go/vms/surveyworker.go
package vms

// The survey worker — the sixth subsystem's worker, and the only one whose input is an archive we do not own.
//
// It watches a camera's DEVICE archive — the card, the NVR — forever, and copies nothing. The video stays
// where it is; what comes out is events:
//
//	survey/watches/<name>                             the unit: cam, kind, params — written by the console
//	<archive>/survey/<name>/e<epoch>/…events.jsonl    what the model saw
//	<archive>/survey/<name>/frontier.json             how far it has watched
//
// Three things shape it, and all three come from the archive being somebody else's.
//
// The INDEX says where to look. A device recording on motion has mostly nothing between its first and last
// minute, and walking that as if it were continuous would spend the whole budget on silence.
//
// The DOOR is the limit, not the GPU. A device allows two playback sessions and they are shared with the
// operator watching and the recorder saving; when they are gone this worker waits and says so, rather than
// retrying into a wall.
//
// And the frontier MOVES ON. It is not a plan to finish but an edge to keep up with — so what matters is not
// "did we cover everything" but how far behind the device's newest minute we are, which is what `lag` in the
// heartbeat says.

import (
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"

	p "vmsserver/w2cplatform"
)

// SURVEY is the subsystem's name — the keys its slot, epoch, assignment and heartbeat live under.
var SURVEY = SurveySpec.Sub()

// Fetch is one read through the holder's door: what costs a session. ErrDeviceBusy when both are in use.
type Fetch func(url string, t0, t1 float64) error

// FetchBytes: the door over HTTP. The bytes are not kept: this subsystem copies nothing, and what it took
// the session for is the decoding, not the file.
func FetchBytes(url string, t0, t1 float64) error {
	if url == "" {
		return fmt.Errorf("no playback url: %w", ErrDeviceBusy)
	}
	c := http.Client{Timeout: 30 * time.Second}
	resp, err := c.Get(url + "?from=" + ftoa(t0) + "&to=" + ftoa(t1))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	if resp.StatusCode == 503 {
		return fmt.Errorf("%s: %w", url, ErrDeviceBusy) // the device's ceiling, not ours
	}
	if resp.StatusCode != 200 {
		return fmt.Errorf("%s: %s", url, resp.Status)
	}
	return nil
}

type SurveyWorker struct {
	*p.Worker
	Models      map[string]func(p.Row) Model
	Capacity    int
	Server      string
	Labels      []string
	ArchiveRoot string
	// SecondsPerPass: seconds of MEDIA one pass may consume per watch. The same argument as the scan's budget
	// of stretches: a pass that ran until it caught up is a worker that stops heartbeating while it does.
	SecondsPerPass float64
	Step           float64 // seconds of media between two looks
	// Settle: never closer to the device's newest minute than this — it is still being written.
	Settle float64
	Fetch  Fetch // the door: reading, and what costs a session
	Index  Index // the listing: where the footage is, and it costs none

	// Hits: `<cam>|<from>|<to>` a model liked — the console asks the recorder. The same window as the
	// recorder's `closed`, and for the same reason: a heartbeat is one object under a ceiling, and the
	// console's decision is idempotent on its own because the request's id is the range.
	Hits          []string
	EventsWritten int

	running      map[string]Model
	statusByUnit map[string]map[string]any
}

// HitsReported: how many kept stretches the heartbeat carries.
const HitsReported = 32

type SurveyOptions struct {
	Models         map[string]func(p.Row) Model
	Capacity       int
	Server         string
	Labels         []string
	ArchiveRoot    string
	SecondsPerPass float64
	Step           float64
	Fetch          Fetch
	Index          Index
	Worker         p.WorkerOptions
}

func NewSurveyWorker(name string, vars p.Variables, objects p.ObjectStore, o SurveyOptions) (*SurveyWorker, error) {
	wo := o.Worker
	wo.Name = ""
	w := p.NewWorker(SURVEY, vars, objects, wo)
	if _, err := w.ClaimSlot(slotOr(name, "s")); err != nil {
		return nil, err
	}
	s := &SurveyWorker{Worker: w, Models: o.Models, Capacity: o.Capacity, Server: o.Server, Labels: o.Labels,
		ArchiveRoot: o.ArchiveRoot, SecondsPerPass: o.SecondsPerPass, Step: o.Step, Settle: 30,
		Fetch: o.Fetch, Index: o.Index, running: map[string]Model{}, statusByUnit: map[string]map[string]any{}}
	if s.Models == nil {
		s.Models = map[string]func(p.Row) Model{"motion": NewFakeModel, "linecross": NewFakeModel, "lpr": NewFakeModel}
	}
	if s.Capacity == 0 {
		s.Capacity = 2
	}
	if s.Labels == nil {
		s.Labels = []string{"gpu"}
	}
	if s.ArchiveRoot == "" {
		s.ArchiveRoot = "/data/archive"
	}
	if s.Server == "" {
		s.Server, _ = os.Hostname()
	}
	if s.SecondsPerPass == 0 {
		s.SecondsPerPass = 600
	}
	if s.Step == 0 {
		s.Step = 1
	}
	if s.Fetch == nil {
		s.Fetch = FetchBytes
	}
	if s.Index == nil {
		s.Index = DeviceRecordings
	}
	return s, nil
}

func (s *SurveyWorker) watchRow(unit string) p.Row {
	items, _, _ := s.Vars.Get(SURVEY.Config("watches", unit))
	if items == nil || items["deleted"] == "true" {
		return nil
	}
	return SurveySpec.RowOf(items)
}

// device: where the device's footage is and where to read it — (index_url, playback_url, oldest, newest),
// all from the HOLDER's heartbeat, which is how everything here is found.
func (s *SurveyWorker) device(cam string) (indexURL, playURL string, oldest, newest float64, ok bool) {
	h, found := p.HolderOf(s.Objects, "vms/", cam, s.Wall(), p.HolderQuery{Field: "coverage"})
	if !found {
		return "", "", 0, 0, false
	}
	cov, isMap := h.Status["coverage"].(map[string]any)
	idx, _ := h.Status["index_url"].(string)
	if !isMap || idx == "" {
		return "", "", 0, 0, false
	}
	play, _ := h.Status["playback_url"].(string)
	return idx, play, p.ToFloat(cov["from"]), p.ToFloat(cov["to"]), true
}

// ReconcileOnce: advance every watch by at most a budget of media seconds.
func (s *SurveyWorker) ReconcileOnce() []string {
	wanted := map[string]bool{}
	units := append([]string{}, s.Assignment().Units...)
	sort.Strings(units)
	for _, unit := range units {
		wanted[unit] = true
		row := s.watchRow(unit)
		if row == nil {
			continue
		}
		if !row.Bool("enabled") {
			s.stop(unit)
			s.statusByUnit[unit] = s.status(unit, row, "pending", "", nil, 0)
			continue
		}
		factory, known := s.Models[p.Str(row["kind"])]
		if !known {
			s.statusByUnit[unit] = s.status(unit, row, "unsupported", "", nil, 0)
			continue
		}
		cam := p.Str(row["cam"])
		indexURL, playURL, oldest, newest, ok := s.device(cam)
		if !ok {
			s.stop(unit)
			s.statusByUnit[unit] = s.status(unit, row, "waiting", "nobody holds this camera, or its device has no archive", nil, 0)
			continue
		}
		front := NewFrontier(s.ArchiveRoot, unit)
		at, started := front.Read()
		if !started {
			// The first time. `earliest` means thirty days of backlog on the day somebody enables it, which
			// is a decision the row makes and not a default this code picks.
			at = newest
			if p.Str(row["start"]) == "earliest" {
				at = oldest
			}
		}
		edge := newest - s.Settle // the last minutes are being written; reading them gets a torn end
		if at >= edge {
			s.statusByUnit[unit] = s.status(unit, row, "running", "", front, newest)
			continue
		}
		want := [2]float64{at, min(at+s.SecondsPerPass, edge)}
		spans, err := s.Index(indexURL, want[0], want[1])
		switch {
		case errors.Is(err, ErrCannotList):
			spans = [][2]float64{want} // the summary is all there is: watch the window as if it were continuous
		case err != nil:
			s.statusByUnit[unit] = s.status(unit, row, "waiting", "the holder is not answering", front, newest)
			continue
		}

		if _, have := s.Epochs[unit]; !have {
			if _, err := s.TakeEpoch(unit); err != nil { // one writer of survey/<unit>/… at a time
				continue
			}
		}
		model, up := s.running[unit]
		if !up {
			model = factory(row)
			s.running[unit] = model
		}
		if !s.MayWrite(unit) {
			continue
		}

		var busy error
		var fired []float64
		var watched [][2]float64
		camN, _ := strconv.Atoi(cam)
		for _, sp := range spans {
			if err := s.Fetch(playURL, sp[0], sp[1]); err != nil { // the door: this is what costs a session
				busy = err // either way nothing past here was watched, and the frontier must not claim it
				break
			}
			watched = append(watched, sp)
			for ts := sp[0]; ts < sp[1]; ts += s.Step {
				for _, obs := range model.Observe(ts) {
					fields := map[string]any{"cam": camN, "watch": unit, "source": "device"}
					for k, v := range obs.Fields {
						if k != "cam" {
							fields[k] = v
						}
					}
					if _, err := p.NewEventLog(s.ArchiveRoot, SURVEY.Name, unit, s.Epochs[unit], 0).Append(ts, obs.Kind, fields); err != nil {
						continue
					}
					s.EventsWritten++
					fired = append(fired, ts)
				}
			}
		}
		// The frontier moves over the whole window, not from span to span. A gap in the device's own recording
		// is nothing to watch and nothing to come back for — leaving the frontier at its edge would park the
		// survey in front of every quiet night for ever. When the door closed half way, it moves to the end
		// of what WAS watched: those events are written, and watching them again would write them twice.
		through, moved := want[1], true
		if busy != nil {
			moved = len(watched) > 0
			if moved {
				through = watched[len(watched)-1][1]
			}
		}
		if moved {
			front.Set(through)
			if p.Str(row["keep"]) == "hits" && len(fired) > 0 {
				// Watch everything, copy what a model liked. The stretches are reported, not written: a
				// worker's token writes no configuration, and the row that asks the recorder for a range is
				// configuration (`rec/requests/<id>`). The console turns these into requests.
				for _, h := range HitSpans(fired, watched, p.ToFloat(row["pre"]), p.ToFloat(row["post"]), p.ToFloat(row["join"])) {
					s.Hits = tail(append(s.Hits, cam+"|"+strconv.FormatFloat(h[0], 'f', 0, 64)+"|"+strconv.FormatFloat(h[1], 'f', 0, 64)), HitsReported)
				}
			}
		}
		if busy != nil {
			// Not a failure and not a retry: the two sessions belong to the operator watching this gap and to
			// the recorder saving it, and a survey is the one of the three that can wait.
			why := "the device has no free session"
			if !errors.Is(busy, ErrDeviceBusy) {
				why = "the holder is not answering: " + busy.Error()
			}
			s.statusByUnit[unit] = s.status(unit, row, "waiting", why, front, newest)
			continue
		}
		s.statusByUnit[unit] = s.status(unit, row, "running", "", front, newest)
	}
	for unit := range s.running {
		if !wanted[unit] {
			s.stop(unit)
			s.Release(unit)
		}
	}
	for unit := range s.statusByUnit {
		if !wanted[unit] {
			delete(s.statusByUnit, unit)
		}
	}
	s.RenewLeases()
	out := make([]string, 0, len(s.running))
	for u := range s.running {
		out = append(out, u)
	}
	sort.Strings(out)
	return out
}

// status: `lag` is the answer this subsystem exists to give about itself — how far behind the device's
// newest minute we are. A survey that cannot keep up is not broken and not finished: it is behind, and the
// only way anyone finds out is if it says so.
func (s *SurveyWorker) status(unit string, row p.Row, phase, why string, front *Frontier, newest float64) map[string]any {
	st := map[string]any{"id": unit, "cam": p.Str(row["cam"]), "kind": p.Str(row["kind"]), "phase": phase}
	if why != "" {
		st["why"] = why
	}
	if front != nil {
		at, ok := front.Read()
		st["watched_through"] = at
		if ok {
			st["lag"] = max(0, newest-at)
		}
	}
	st["events"] = s.EventsWritten
	return st
}

func (s *SurveyWorker) stop(unit string) {
	if m, ok := s.running[unit]; ok {
		m.Close()
		delete(s.running, unit)
	}
}

func (s *SurveyWorker) Headroom() int {
	if n := s.Capacity - len(s.running); n > 0 {
		return n
	}
	return 0
}

func (s *SurveyWorker) HeartbeatOnce() error {
	names := make([]string, 0, len(s.statusByUnit))
	for u := range s.statusByUnit {
		names = append(names, u)
	}
	sort.Strings(names)
	status := make([]map[string]any, 0, len(names))
	for _, u := range names {
		status = append(status, s.statusByUnit[u])
	}
	return s.HeartbeatWith(status, map[string]any{
		"server": s.Server, "instance": s.Instance, "labels": strings.Join(s.Labels, ","),
		"capacity": s.Capacity, "headroom": s.Headroom(), "conflicts": s.Conflicts(), "events": s.EventsWritten,
		"hits": strings.Join(s.Hits, ","),
	})
}

// Status is what the tests read: the per-unit entry this worker would publish.
func (s *SurveyWorker) Status(unit string) map[string]any { return s.statusByUnit[unit] }

File: vmsserver/vms/surveyworker.py
"""The survey worker — the sixth subsystem's worker, and the only one whose input
is an archive we do not own.

It watches a camera's DEVICE archive — the card, the NVR — forever, and copies
nothing. The video stays where it is; what comes out is events:

    survey/watches/<name>        the unit: cam, kind, params — written by the console
    survey/workers/<s>           the assignment, written by the survey controller
    survey/<s>/heartbeat         capacity, headroom, labels; per-unit status: phase, lag, events
    <archive>/survey/<name>/e<epoch>/…events.jsonl   what the model saw
    <archive>/survey/<name>/frontier.json            how far it has watched

Three things shape it, and all three come from the archive being somebody else's.

The INDEX says where to look. A device recording on motion has mostly nothing
between its first and last minute, and walking that as if it were continuous
would spend the whole budget on silence.

The DOOR is the limit, not the GPU. A device allows two playback sessions and
they are shared with the operator watching and the recorder saving; when they
are gone this worker waits and says so, rather than retrying into a wall.

And the frontier MOVES ON. It is not a plan to finish but an edge to keep up
with — so what matters is not "did we cover everything" but how far behind the
device's newest minute we are, which is what `lag` in the heartbeat says.
"""
from __future__ import annotations

import logging
import os
import time
import urllib.error

from w2cplatform import runtime
from w2cplatform.console import holder_of
from w2cplatform.contract import Worker
from w2cplatform.events import EventLog
from w2cplatform.variables import Variables

from .config import SURVEY_SPEC
from .detworker import FakeModel
from .scan import SURVEY, Frontier, device_recordings, hit_spans

SURVEY_SUB = SURVEY_SPEC.sub
log = logging.getLogger("vms.surveyworker")


class SurveyWorker(Worker):
    """`name` is a slot (`s-1`); `models` maps a kind to a factory `(row) -> Model`,
    the same registry the live detector and the scan use."""

    # Seconds of MEDIA one pass may consume per watch. The same argument as the scan's budget of stretches
    # (Lesson 21): a pass that ran until it caught up is a worker that stops heartbeating while it does.
    SECONDS_PER_PASS = 600.0
    STEP = 1.0                     # seconds of media between two looks
    SETTLE = 30.0                  # never closer to the device's newest minute than this: it is still being written
    # How many kept stretches the heartbeat carries. The same window as the recorder's `closed` (Lesson 22)
    # and for the same reason: a heartbeat is one object under a ceiling, and the console's decision is
    # idempotent on its own because the request's id is the range.
    HITS_REPORTED = 32

    def __init__(self, name: str | None, vars_: Variables, objects, models: dict | None = None,
                 capacity: int | None = None, clock=time.monotonic, wall=time.time, server: str | None = None,
                 archive_root: str | None = None, env: dict | None = None, step: float | None = None,
                 fetch=None, index=None):
        env = dict(os.environ if env is None else env)
        super().__init__(SURVEY_SUB, None, vars_, objects, clock=clock, wall=wall)
        self.claim_slot(prefer=name if name is not None else runtime.slot(env, "SURVEY_NAME", "s"))
        self.models = models if models is not None else {"motion": FakeModel, "linecross": FakeModel, "lpr": FakeModel}
        self.capacity = capacity if capacity is not None else int(env.get("SURVEY_CAPACITY", "2"))
        self.server = runtime.server(env, server)
        self.labels = runtime.labels(env, "gpu")
        self.archive_root = archive_root or env.get("ARCHIVE", "/data/archive")
        self.step = self.STEP if step is None else float(step)
        self.fetch = fetch or _fetch_bytes           # the door: reading, and what costs a session
        self.index = index or device_recordings      # the listing: where the footage is, and it costs none
        self.running: dict[str, object] = {}
        self.status_by_unit: dict[str, dict] = {}
        self.events_written = 0
        self.hits: list[str] = []                    # `<cam>|<from>|<to>` a model liked — the console asks the recorder

    def watch_row(self, unit: str) -> dict | None:
        it, _ = self.vars.get(SURVEY_SUB.config("watches", unit))
        return SURVEY_SPEC.row(it) if it and it.get("deleted") != "true" else None

    # Where the device's footage is and where to read it: `(index_url, playback_url, oldest, newest)`, all
    # from the HOLDER's heartbeat, which is how everything here is found.
    def device(self, cam) -> tuple[str, str, float, float] | None:
        found = holder_of(self.objects, "vms/", str(cam), self.wall(), field="coverage")
        if found is None or not found[2].get("coverage") or not found[2].get("index_url"):
            return None
        cov = found[2]["coverage"]
        return found[2]["index_url"], found[2].get("playback_url", ""), float(cov["from"]), float(cov["to"])

    # -- the pass: advance every watch by at most a budget of media seconds ----------------------------
    def reconcile_once(self, now: float | None = None) -> list[str]:
        now = self.wall() if now is None else now
        wanted = set(self.assignment().units)
        for unit in sorted(wanted):
            row = self.watch_row(unit)
            if row is None:
                continue
            if not row["enabled"]:
                self._stop(unit); self.status_by_unit[unit] = self._status(unit, row, "pending"); continue
            if row["kind"] not in self.models:
                self.status_by_unit[unit] = self._status(unit, row, "unsupported"); continue

            dev = self.device(row["cam"])
            if dev is None:
                self._stop(unit)
                self.status_by_unit[unit] = self._status(unit, row, "waiting",
                                                         why="nobody holds this camera, or its device has no archive")
                continue
            index_url, play_url, oldest, newest = dev
            front = Frontier(self.archive_root, unit)
            at = front.read()
            if at is None:
                # The first time. `earliest` means thirty days of backlog on the day somebody enables it,
                # which is a decision the row makes and not a default this code picks.
                at = oldest if row["start"] == "earliest" else newest

            edge = newest - self.SETTLE          # the last minutes are being written; reading them gets a torn end
            if at >= edge:
                self.status_by_unit[unit] = self._status(unit, row, "running", front=front, newest=newest)
                continue

            want = (at, min(at + self.SECONDS_PER_PASS, edge))
            try:
                spans = self.index(index_url, want[0], want[1])
            except Exception:                            # noqa: BLE001 — the holder is there and not answering
                self.status_by_unit[unit] = self._status(unit, row, "waiting", why="the holder is not answering",
                                                         front=front, newest=newest)
                continue

            if unit not in self.epochs:
                self.take_epoch(unit)                    # one writer of survey/<unit>/… at a time
            model = self.running.get(unit)
            if model is None:
                model = self.running[unit] = self.models[row["kind"]](row)
            if not self.may_write(unit):
                continue

            busy, fired, watched = False, [], []
            for a, b in (spans if spans is not None else [want]):
                try:
                    self.fetch(play_url, a, b)           # the door: this is what costs a session
                except _Busy:
                    busy = True
                    break
                watched.append((a, b))
                for ts, kind, fields in self._watch(model, a, b):
                    EventLog(self.archive_root, SURVEY, unit, self.epochs[unit]).append(
                        ts, kind, cam=int(row["cam"]), watch=unit, source="device", **fields)
                    self.events_written += 1
                    fired.append(ts)
            # The frontier moves over the whole window, not from span to span. A gap in the device's own
            # recording is nothing to watch and nothing to come back for — leaving the frontier at its edge
            # would park the survey in front of every quiet night for ever. When the door closed half way, it
            # moves to the end of what WAS watched: those events are written, and watching them again would
            # write them twice.
            through = want[1] if not busy else (watched[-1][1] if watched else None)
            if through is not None:
                front.set(through)
                if row["keep"] == "hits" and fired:
                    # Watch everything, copy what a model liked. The stretches are reported, not written: a
                    # worker's token writes no configuration, and the row that asks the recorder for a range
                    # is configuration (`rec/requests/<id>` — Lesson 21). The console turns these into requests.
                    for a, b in hit_spans(fired, watched, row["pre"], row["post"], row["join"]):
                        self.hits = (self.hits + [f"{row['cam']}|{a:.0f}|{b:.0f}"])[-self.HITS_REPORTED:]
            if busy:
                # Not a failure and not a retry: the two sessions belong to the operator watching this gap and
                # to the recorder saving it, and a survey is the one of the three that can wait.
                self.status_by_unit[unit] = self._status(unit, row, "waiting", why="the device has no free session",
                                                         front=front, newest=newest)
                continue
            self.status_by_unit[unit] = self._status(unit, row, "running", front=front, newest=newest)

        for unit in list(self.running):
            if unit not in wanted:
                self._stop(unit); self.release(unit)
        for unit in list(self.status_by_unit):
            if unit not in wanted:
                self.status_by_unit.pop(unit, None)
        self.renew_leases()
        return sorted(self.running)

    def _watch(self, model, a: float, b: float):
        ts = a
        while ts < b:
            for kind, fields in model.observe(ts):
                yield ts, kind, fields
            ts += self.step

    # `lag` is the answer this subsystem exists to give about itself: how far behind the device's newest
    # minute we are. A survey that cannot keep up is not broken and not finished — it is behind, and the
    # only way anyone finds out is if it says so.
    def _status(self, unit: str, row: dict, phase: str, why: str = "", front=None, newest: float | None = None) -> dict:
        st = {"id": unit, "cam": row["cam"], "kind": row["kind"], "phase": phase}
        if why:
            st["why"] = why
        if front is not None:
            at = front.read()
            st["watched_through"] = at if at is not None else 0.0
            if newest is not None and at is not None:
                st["lag"] = max(0.0, newest - at)
        st["events"] = self.events_written
        return st

    def _stop(self, unit: str) -> None:
        m = self.running.pop(unit, None)
        if m is not None:
            m.close()

    def headroom(self) -> int:
        return max(0, self.capacity - len(self.running))

    def heartbeat_once(self) -> None:
        self.heartbeat(list(self.status_by_unit.values()), server=self.server, instance=self.instance,
                       labels=",".join(self.labels), capacity=self.capacity, headroom=self.headroom(),
                       conflicts=self.conflicts(), events=self.events_written, hits=",".join(self.hits))

    def run(self, poll: float = 2.0, stop=None) -> None:
        import threading
        stop = stop or threading.Event()
        while not stop.is_set():
            try:
                self.reconcile_once(); self.heartbeat_once()
            except Exception:                            # noqa: BLE001
                log.exception("survey pass failed")
            stop.wait(poll)
        for unit in list(self.running):
            self._stop(unit)
        self.release_slot()


class _Busy(Exception):
    """The device's two playback sessions are both in use."""


# One read through the holder's door. The bytes are not kept: this subsystem copies nothing, and what it
# took the session for is the decoding, not the file.
def _fetch_bytes(url: str, t0: float, t1: float, timeout: float = 30.0) -> bytes:
    import urllib.request
    if not url:
        raise _Busy("no playback url")
    try:
        with urllib.request.urlopen(f"{url}?from={t0}&to={t1}", timeout=timeout) as r:
            return r.read()
    except urllib.error.HTTPError as e:
        if e.code == 503:
            raise _Busy(str(e)) from e                   # the device's ceiling, not ours
        raise
